[DTensor][random] defer DTensor RNG state sync until first random op call or manual_seed call; support more flexible OffsetBasedRNGTracker init

XilunWu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #147025

Resolves https://github.com/pytorch/pytorch/issues/146767.

May also resolve https://github.com/pytorch/pytorch/issues/147584.

### Summary
This PR removes the RNG tracker init from the `distribute_tensor` call for the following reasons:

1. if the user does not use random ops on DTensor, there's no need to init DTensor RNG which currently requires CUDA device to be present. 
2. this complies with the 0-communication semantic of `src_data_rank=None` shard distribution.

Besides, `OffsetBasedRNGTracker` only accepts `DeviceMesh` argument to its constructor method.

### Consequence

DTensor RNG initialization is delayed till the first DTensor random ops call or `torch.distributed.tensor.random.manual_seed`.

### Test
`pytest test/distributed/tensor/test_random_ops.py`
`pytest test/distributed/tensor/parallel/test_tp_random_state.py`
`pytest test/distributed/tensor/parallel/test_tp_style.py`

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

Differential Revision: [D70201856](https://our.internmc.facebook.com/intern/diff/D70201856)

XilunWu committed:
[DTensor][random] defer DTensor RNG state sync until first random op call or manual_seed call; support more flexible OffsetBasedRNGTracker init

[ghstack-poisoned]

kwen2501 commented on torch/distributed/tensor/_random.py:
```diff
@@ -161,9 +162,15 @@ class OffsetBasedRNGTracker(_RNGStateTracker):
-        rng_state = self._device_handle.get_rng_state().to(device_type)
+    def __init__(
+        self,
+        device: Union[int, str, torch.device] = "cuda",
```
nit: remove "cuda" ?

kwen2501 commented on torch/distributed/tensor/_dispatch.py:
```diff
@@ -198,7 +198,9 @@ def dispatch(
                     # did not already construct one
-                    random._rng_tracker = random.OffsetBasedRNGTracker(mesh.device_type)
+                    random._rng_tracker = random.OffsetBasedRNGTracker(
+                        mesh.device_type, mesh
```
nit: seems like we can consolidate these two args to `mesh` only?

kwen2501 commented on torch/distributed/tensor/_random.py:
```diff
@@ -185,7 +192,10 @@ def _distribute_region(self, spec: DTensorSpec):
             self._set_pre_op_offset(spec)
-            with torch.random.fork_rng(self._devices, device_type=self._device_type):
+            with torch.random.fork_rng(
+                devices=[self._device], device_type=self._device.type
```
nit: consolidate arguments too?

kwen2501 commented on torch/distributed/tensor/_random.py:
```diff
@@ -366,3 +376,24 @@ def _calc_shard_linear_idx(
+
+
+def _resolve_device(
+    device: Union[int, str, torch.device] = "cuda",
```
nit: remove "cuda"

kwen2501 commented on torch/distributed/tensor/_random.py:
```diff
@@ -366,3 +376,24 @@ def _calc_shard_linear_idx(
+) -> torch.device:
+    if isinstance(device, torch.device):
+        return device
+    elif isinstance(device, int):
+        return torch.device(f"cuda:{device}")
```
Would the use of "cuda" seem arbitrary? Can we get the type from `device_mesh`?

kwen2501 reviewed (approved):
LGTM in general. I left some comments.

kwen2501 commented:
Can you also add a PR description?

XilunWu committed:
Update on "[DTensor][random] defer DTensor RNG state sync until first random op call or manual_seed call; support more flexible OffsetBasedRNGTracker init"


Resolves https://github.com/pytorch/pytorch/issues/146767.

May also resolve https://github.com/pytorch/pytorch/issues/147584.

### Summary
This PR removes the RNG tracker init from the `distribute_tensor` call for the following reasons:

1. if the user does not use random ops on DTensor, there's no need to init DTensor RNG which currently requires CUDA device to be present. 
2. this complies with the 0-communication semantic of `src_data_rank=None` shard distribution.

Besides, `OffsetBasedRNGTracker` only accepts `DeviceMesh` argument to its constructor method.

### Consequence

DTensor RNG initialization is delayed till the first DTensor random ops call or `torch.distributed.tensor.random.manual_seed`.

### Test
`pytest test/distributed/tensor/test_random_ops.py`
`pytest test/distributed/tensor/parallel/test_tp_random_state.py`
`pytest test/distributed/tensor/parallel/test_tp_style.py`

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k 

[ghstack-poisoned]

XilunWu committed:
Update on "[DTensor][random] defer DTensor RNG state sync until first random op call or manual_seed call; support more flexible OffsetBasedRNGTracker init"


Resolves https://github.com/pytorch/pytorch/issues/146767.

May also resolve https://github.com/pytorch/pytorch/issues/147584.

### Summary
This PR removes the RNG tracker init from the `distribute_tensor` call for the following reasons:

1. if the user does not use random ops on DTensor, there's no need to init DTensor RNG which currently requires CUDA device to be present. 
2. this complies with the 0-communication semantic of `src_data_rank=None` shard distribution.

Besides, `OffsetBasedRNGTracker` only accepts `DeviceMesh` argument to its constructor method.

### Consequence

DTensor RNG initialization is delayed till the first DTensor random ops call or `torch.distributed.tensor.random.manual_seed`.

### Test
`pytest test/distributed/tensor/test_random_ops.py`
`pytest test/distributed/tensor/parallel/test_tp_random_state.py`
`pytest test/distributed/tensor/parallel/test_tp_style.py`

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k 

[ghstack-poisoned]

XilunWu commented:
@XilunWu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D70201856).

wconstab commented on torch/distributed/tensor/_random.py:
```diff
@@ -159,11 +158,25 @@ class OffsetBasedRNGTracker(_RNGStateTracker):
+        if self._device.type != "cuda":
+            raise RuntimeError(
+                f"{self.__class__.__name__} instantiation requires the presence of "
+                f"CUDA/CUDA-like device. Got {self._device.type} instead."
```
PR Lgtm overall. 
Remind me if we have a cpu-compatible rng tracker today?

XilunWu replied:
@wconstab unfortunately no, and it may not be very useful IMO, because the main usage of DTensor random ops is model init and `dropout()` in model training. Neither is expected to run on CPU. 

If we see a use case, it's also possible to achieve a CPU RNG tracker by either us or users. Users can also extend the base class to make a tracker for CPU. Let me know if we already have such a use case.

XilunWu commented:
@pytorchbot merge -f "unrelated CI failure -- [periodic / linux-focal-cuda12.4-py3.10-gcc9-bazel-test / build-and-test (default, 1, 1, lf.linux.4xlarge.nvidia.gpu)](https://hud.pytorch.org/pr/pytorch/pytorch/147025#37813547410) is on main"
